### PR TITLE
CSI raw block volume support

### PR DESCRIPTION
**Pull request by BaiHuoYu**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
CSI raw block volume support
**Special notes for your reviewer**:

**Release note**:

```release-note
CSI raw block volume support
```

### CODE REVIEW

File: csi/server/plugin/opensds/node.go [hunk 1]
```diff
@@ -54,17 +55,37 @@ func getVolumeAndAttachment(volumeId string, attachmentId string) (*model.Volume
...
-		_, err = exec.Command("mount", device, mountpoint).CombinedOutput()
-	}
+		_, err = os.Lstat(mountpoint)
 
-	if nil != err {
-		return status.Error(codes.Aborted, fmt.Sprintf("failed to mount: %v", err.Error()))
+		if err != nil && os.IsNotExist(err) {
+			glog.V(5).Infof("Mountpoint=%v is not exist", mountpoint)
+		} else {
+			glog.Errorf("Mountpoint=%v already exists!", mountpoint)
```
Comment (xing-yang) [inline csi/server/plugin/opensds/node.go]: error message should start with lower case letter and without ending punctuation.

**Comment by leonwanghui** [comment]
Please change the base branch to `development`

**Comment by xing-yang** [comment]
@BaiHuoYu Please address Leon's comments.

File: csi/server/plugin/opensds/node.go [hunk 2]
```diff
@@ -54,17 +55,37 @@ func getVolumeAndAttachment(volumeId string, attachmentId string) (*model.Volume
...
-	}
+		_, err = os.Lstat(mountpoint)
 
-	if nil != err {
-		return status.Error(codes.Aborted, fmt.Sprintf("failed to mount: %v", err.Error()))
+		if err != nil && os.IsNotExist(err) {
+			glog.V(5).Infof("Mountpoint=%v does not exist", mountpoint)
+		} else {
+			glog.Errorf("mountpoint=%v already exists", mountpoint)
+			_, err := exec.Command("rm", "-rf", mountpoint).CombinedOutput()
```
Comment (xing-yang) [inline csi/server/plugin/opensds/node.go]: If mount point already exists, is it safe to remove it here?  Is this to handle idempotency? In any case, please add comments to explain why this is needed and why this is safe to do.

File: csi/server/plugin/opensds/node.go [hunk 3]
```diff
@@ -54,17 +55,37 @@ func getVolumeAndAttachment(volumeId string, attachmentId string) (*model.Volume
...
+
+			if nil != err {
+				return err
```
Comment (xing-yang) [inline csi/server/plugin/opensds/node.go]: Log an error msg.

File: csi/server/plugin/opensds/node.go [hunk 4]
```diff
@@ -54,17 +55,37 @@ func getVolumeAndAttachment(volumeId string, attachmentId string) (*model.Volume
...
+			}
+		}
+
+		err = os.Symlink(device, mountpoint)
+		if err != nil {
+			return err
```
Comment (xing-yang) [inline csi/server/plugin/opensds/node.go]: Log an error msg.

File: csi/server/plugin/opensds/node.go [hunk 5]
```diff
@@ -242,49 +263,66 @@ func (p *Plugin) NodeStageVolume(
...
-	if "" == curFSType {
-		_, err := exec.Command("mkfs", "-t", hopeFSType, "-F", device).CombinedOutput()
-		if err != nil {
-			return nil, status.Error(codes.Aborted, fmt.Sprintf("failed to mkfs: %v", err.Error()))
+	var mountFlags []string
+	if nil == block {
+		vol.Metadata[KCSIVolumeMode] = "Filesystem"
+		mountFlags = mnt.MountFlags
+		_, err = exec.Command("findmnt", device, mountpoint).CombinedOutput()
+		glog.V(5).Infof("findmnt err: %v \n", err)
```
Comment (xing-yang) [inline csi/server/plugin/opensds/node.go]: We are not sure if there is an error or not to log here.  If this is for debugging, I think you should add "device" and "mountpoint" in the log msg as well.

File: csi/server/plugin/opensds/node.go [hunk 6]
```diff
@@ -312,15 +350,23 @@ func (p *Plugin) NodeUnstageVolume(
...
 		return nil, err
 	}
 
-	vol, attachment, err := getVolumeAndAttachmentByVolumeId(req.VolumeId)
-	if err != nil {
-		return nil, err
+	if KCSIFilesystem == vol.Metadata[KCSIVolumeMode] {
+		err = connector.Umount(req.StagingTargetPath)
+		if err != nil {
+			return nil, err
```
Comment (xing-yang) [inline csi/server/plugin/opensds/node.go]: Log an error msg.

File: csi/server/plugin/opensds/node.go [hunk 7]
```diff
@@ -312,15 +350,23 @@ func (p *Plugin) NodeUnstageVolume(
...
+		}
+	}
+
+	if KCSIBlock == vol.Metadata[KCSIVolumeMode] {
+		_, err = exec.Command("rm", "-rf", req.StagingTargetPath).CombinedOutput()
+		if err != nil {
+			return nil, err
```
Comment (xing-yang) [inline csi/server/plugin/opensds/node.go]: Log an error msg

File: csi/server/plugin/opensds/node.go [hunk 8]
```diff
@@ -412,15 +467,23 @@ func (p *Plugin) NodeUnpublishVolume(
...
+		err := connector.Umount(req.TargetPath)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if KCSIBlock == vol.Metadata[KCSIVolumeMode] {
+		_, err = exec.Command("rm", "-rf", req.TargetPath).CombinedOutput()
+		if err != nil {
+			return nil, err
```
Comment (xing-yang) [inline csi/server/plugin/opensds/node.go]: Log an error msg

File: csi/server/plugin/opensds/node.go (hunk 2 above)
Comment (BaiHuoYu) [reply csi/server/plugin/opensds/node.go]: The mountpoint deleted here is a folder or a soft connection. From the test results, this is fine.

File: csi/server/plugin/opensds/node.go (hunk 5 above)
Comment (BaiHuoYu) [reply csi/server/plugin/opensds/node.go]: Returning err here is not an error, it just means that the volume has been published.

**Comment by xing-yang** [comment]
@BaiHuoYu Please take a look of CI failure.

**Pull request closed without merging**